### PR TITLE
Keras-like API functional merge and some fix

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

1. Functional `merge` method in Keras.
Different from `Merge` layer, it merges a list of tensors (NOT layers) in a graph.
2. Refactor Python Keras layer **kwargs accept `name`, `bigdl_type`.
3. Fix some layer arguments to sync with Keras API.
4. Fix a bug in Merge concat and Reshape.
5. Fix some grammar issues in docs.

ResNet50, VGG16, VGG19, InceptionV3 models in Keras (with trained weights) can be written in BigDL use the new API with compatible prediction results.

## How was this patch tested?

Jenkins

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/246/

**Comment by hkvision** [comment]
@jason-dai have a review? https://github.com/intel-analytics/BigDL/pull/2313/files#diff-0df9a9274dd9569e7ac31c276b1fac24
Keras code is here https://github.com/keras-team/keras/blob/1.2.2/keras/applications/resnet50.py

**Comment by hkvision** [comment]
@jason-dai I first removed keras code from this PR. For those Keras models, I ran and compared results with Keras locally. This PR now mainly fixes some issues for new API to support loading these models.

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/250/

**Comment by hkvision** [comment]
@jason-dai @zhichao-li have a review?

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/284/

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 1]
```diff
@@ -144,8 +136,25 @@ class Merge[T: ClassTag](
...
-    model.add(layer)
-    model.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
+    if (layers != null) {
+      val model = TSequential[T]()
+      val parallel = ParallelTable()
+      var i = 0
+      while(i < layers.length) {
+        val tlayer = layers(i) match {
+          case k: KerasLayer[_, _, T] => k.labor
+          case t: AbstractModule[Activity, Activity, T] => t
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: can we safely remove this checking as the mix usage has been disable?

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/MergeSpec.scala [hunk 2]
```diff
@@ -36,6 +36,20 @@ class MergeSpec extends KerasBaseSpec {
     seq.forward(input) should be (input1 + input2)
   }
 
+  "merge method" should "work correctly" in {
```
Comment (zhichao-li) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/MergeSpec.scala]: Let's refer to the python and add more tests here. i.e nested cases, validate the grad and output etc...

Comment (zhichao-li) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/MergeSpec.scala]: or put them into `test_newapi.py`  if it's hard to get the std and running_mean from scala side.

File: pyspark/bigdl/nn/keras/layer.py [hunk 3]
```diff
@@ -58,6 +58,11 @@ def get_output_shape(self):
 
 
 class KerasLayer(Layer):
+    def __init__(self, jvalue, bigdl_type, name, *args):
+        super(Layer, self).__init__(jvalue, bigdl_type, *args)
+        if name:
+            self.set_name(name)
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/layer.py]: we should propagate the name to scala side as well

Comment (zhichao-li) [reply pyspark/bigdl/nn/keras/layer.py]: somehow we can inject the name on the init method of `Layer` instead of change the constructor in each subclass

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/MergeSpec.scala (hunk 2 above)
Comment (hkvision) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/keras/nn/MergeSpec.scala]: Tested on Python side.

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/job/BigDL-PR-Validation-v3/381/

File: docs/docs/KerasStyleAPIGuide/keras-api-scala.md [hunk 4]
```diff
@@ -89,19 +89,36 @@ Parameters:
 * `input`: An input node or an array of input nodes.
 * `output`: An output node or an array of output nodes.
 
+To merge a list of input __nodes__ (__NOT__ layers), following some merge mode in the Functional API:
+```scala
+import com.intel.analytics.bigdl.nn.keras.Merge.merge
+
+merge(inputs, mode = "sum", concat_axis = -1, name = null) // This will return an output NODE.
```
Comment (zhichao-li) [inline docs/docs/KerasStyleAPIGuide/keras-api-scala.md]: remove name = null here?

File: pyspark/bigdl/nn/keras/layer.py [hunk 5]
```diff
@@ -203,47 +240,37 @@ class BatchNormalization(KerasLayer):
...
+            raise ValueError("For BatchNormalization with tf dim ordering, only axis=-1 is supported for now")
+        super(BatchNormalization, self).__init__(None,
                                                  float(epsilon),
                                                  float(momentum),
                                                  beta_init,
                                                  gamma_init,
                                                  dim_ordering,
-                                                 list(input_shape) if input_shape else None)
-
-    def set_running_mean(self, running_mean):
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/layer.py]: why remove these methods?

File: pyspark/bigdl/nn/keras/layer.py [hunk 6]
```diff
@@ -303,13 +351,15 @@ class Flatten(KerasLayer):
 
     # Arguments
     input_shape: A shape tuple, not including batch.
+    name: String to specify the name of the layer. Default is None.
```
Comment (zhichao-li) [inline pyspark/bigdl/nn/keras/layer.py]: remove this? we would generate a default name in scala, not NOne.

File: pyspark/bigdl/nn/keras/layer.py (hunk 5 above)
Comment (hkvision) [reply pyspark/bigdl/nn/keras/layer.py]: These methods will be called in `nn/layer.py`. Remove the duplicate.

File: pyspark/bigdl/nn/keras/layer.py (hunk 6 above)
Comment (hkvision) [reply pyspark/bigdl/nn/keras/layer.py]: I will remove `Default is None`. This doc line is needed so that users can know they can specify this parameter.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 7]
```diff
@@ -52,8 +53,11 @@ class Merge[T: ClassTag](
   require(mergeMode == "sum" || mergeMode == "mul" || mergeMode == "concat" || mergeMode == "ave"
   || mergeMode == "cos" || mergeMode == "dot" || mergeMode == "max",
   s"Invalid merge mode: $mergeMode")
-  require(layers.length >= 2, s"Merge must have at least two input layers " +
-    s"but found ${layers.length}")
+  if (layers != null) {
+    require(layers.length >= 2, s"Merge must take at least two input layers " +
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: ```require(layers != null and  layers.length >=2```

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Reshape.scala [hunk 8]
```diff
@@ -90,7 +90,7 @@ class Reshape[T: ClassTag](
       layer = InferReshape(targetShape)
     }
     else {
-      layer = com.intel.analytics.bigdl.nn.Reshape(targetShape)
+      layer = com.intel.analytics.bigdl.nn.Reshape(targetShape, batchMode = Some(true))
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Reshape.scala]: unitest to cover this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 9]
```diff
@@ -154,14 +161,15 @@ object Merge {
     inputShape: Shape = null,
     layers: Array[AbstractModule[Activity, Activity, T]]): Shape = {
     val batchInputShape = KerasLayer.addBatch(inputShape)
-    val actualInputShape =
+    val actualInputShape = if (layers != null) {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: throw exception here if layers == null?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 10]
```diff
@@ -116,25 +120,15 @@ class Merge[T: ClassTag](
...
-    while(i < layers.length) {
-      val tlayer = layers(i) match {
-        case k: KerasLayer[_, _, T] => k.labor
-        case t: AbstractModule[Activity, Activity, T] => t
-      }
-      parallel.add(tlayer)
-      i += 1
-    }
-    model.add(parallel)
     val seq = TSequential[T]()
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: useless code?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 11]
```diff
@@ -116,25 +120,15 @@ class Merge[T: ClassTag](
...
     val layer = mergeMode match {
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: rename to merge_layer as its unclear together with parameter "layers"

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala [hunk 12]
```diff
@@ -144,8 +138,21 @@ class Merge[T: ClassTag](
...
+      while(i < layers.length) {
+        parallel.add(layers(i).asInstanceOf[KerasLayer[Activity, Activity, T]].labor)
+        i += 1
+      }
+      model.add(parallel)
+      model.add(layer)
+      model.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
+    }
+    else {
+      layer.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
```
Comment (zhichao-li) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: unittest to cover this case that layers == null

Comment (zhichao-li) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: add comment for the usage here.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Reshape.scala (hunk 8 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Reshape.scala]: Added.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala (hunk 10 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: Removed to `dot` or `cos` only.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala (hunk 11 above)
Comment (hkvision) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/Merge.scala]: Renamed.

File: docs/docs/KerasStyleAPIGuide/keras-api-scala.md (hunk 4 above)
Comment (hkvision) [reply docs/docs/KerasStyleAPIGuide/keras-api-scala.md]: Removed.

**Comment by hkvision** [comment]
jenkins passed http://172.168.2.101:8080/view/PR-V(new)/job/BigDL-PR-Validation-v3/382/
